### PR TITLE
Setup state machine

### PR DESCRIPTION
**Pull request by simonrozsival**
Ref #440 

Some issues I would like to consult:

- [x] The transitions are created in the `TogglSyncService`, which is only a wrapper for the `SyncManager` otherwise, and it is exposed in the `ITogglDataSource`. I am not sure if it is a good idea, maybe there should be rather some factory which would just create a `SyncManager`.
- [x] I am not sure which `IScheduler` should be used - now I went with `Scheduler.CurrentThread`.
- [x] for now I put only the `FullSync()` in `App.cs`, ~~but it should be running in some background thread I suppose, so that is one thing to consider.~~ as it as all _observy_, it is already async.
- [x] There is a problem with `SinceParameters` - the object is retrieved from Realm, but is passed to a different thread and as soon as it is accessed, an exception is thrown. `Foundation` should convert the Realm models to Foundation models as soon as it receives them (@willsb suggested  #having a `.Wrap(realmModel)` function).

### CODE REVIEW

File: Toggl.Foundation/TogglSyncService.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            var persistWorkspaces = new PersistWorkspacesState(database);
+            var persistClients = new PersistClientsState(database);
+            var persistProjects = new PersistProjectsState(database);
+            var persistTimeEntries = new PersistTimeEntriesState(database);
+
+            transitions.ConfigureTransition(entryPoint, fetchAllSince.Start);
+            transitions.ConfigureTransition(fetchAllSince.FetchStarted, persistWorkspaces.Start);
+            transitions.ConfigureTransition(persistWorkspaces.FinishedPersisting, persistClients.Start);
+            transitions.ConfigureTransition(persistClients.FinishedPersisting, persistProjects.Start);
+            transitions.ConfigureTransition(persistProjects.FinishedPersisting, persistTimeEntries.Start);
```
Comment (amulware) [inline Toggl.Foundation/TogglSyncService.cs]: Maybe we want extension methods to allow chaining? `.configure(result, start).configure(result, start)...`

Doesn't have to be part of this PR but could be.

File: Toggl.PrimeRadiant.Realm/Repository.cs [hunk 2]
```diff
@@ -30,7 +30,7 @@ public IObservable<IEnumerable<TModel>> BatchUpdate(IEnumerable<TModel> entities
             Ensure.Argument.IsNotNull(entities, nameof(entities));
             Ensure.Argument.IsNotNull(conflictResolution, nameof(conflictResolution));
 
-            return CreateObservable(() => Adapter.BatchUpdate(entities, matchById, conflictResolution));
+            return Observable.Return(Adapter.BatchUpdate(entities, matchById, conflictResolution));
```
Comment (amulware) [inline Toggl.PrimeRadiant.Realm/Repository.cs]: What's this for?

File: Toggl.Foundation/DataSources/TogglDataSource.cs [hunk 3]
```diff
@@ -20,6 +21,7 @@ public TogglDataSource(ITogglDatabase database, ITogglApi api)
             User = new UserDataSource(database.User, api.User);
             Projects = new ProjectsDataSource(database.Projects);
             TimeEntries = new TimeEntriesDataSource(database.IdProvider, database.TimeEntries);
+            SyncService = new TogglSyncService(database, api, Scheduler.CurrentThread);
```
Comment (amulware) [inline Toggl.Foundation/DataSources/TogglDataSource.cs]: Why this scheduler? Won't that force the state machine to run on whatever thread runs this code? That sounds bad. I figure it should just use the generic default scheduler instead.

File: Toggl.Foundation/TogglSyncService.cs (hunk 1 above)
Comment (simonrozsival) [reply Toggl.Foundation/TogglSyncService.cs]: We could refactor it later, this is quite verbose. On the other hand this is not something you would write very often and is not that important. But that is up to you.

File: Toggl.PrimeRadiant.Realm/Repository.cs (hunk 2 above)
Comment (simonrozsival) [reply Toggl.PrimeRadiant.Realm/Repository.cs]: The first thing didn't work. I didn't investigate this very much yesterday. I can check later if it is unnecessary, when the whole thing works.

File: Toggl.Foundation/DataSources/TogglDataSource.cs (hunk 3 above)
Comment (simonrozsival) [reply Toggl.Foundation/DataSources/TogglDataSource.cs]: As I wrote in the description of the PR, some help here would be much appreciated.

File: Toggl.PrimeRadiant.Realm/Repository.cs (hunk 2 above)
Comment (simonrozsival) [reply Toggl.PrimeRadiant.Realm/Repository.cs]: It really doesn't work without this change.

Comment (simonrozsival) [reply Toggl.PrimeRadiant.Realm/Repository.cs]: The problem is that we don’t ever subscribe to the observable returned by the Repository.BatchUpdate, so the func is never executed.

Comment (amulware) [reply Toggl.PrimeRadiant.Realm/Repository.cs]: Do we want the BatchUpdate observable to be hot though? that means that we may or may not get the returned values when we do want them. What is the desired behaviour here?

Comment (simonrozsival) [reply Toggl.PrimeRadiant.Realm/Repository.cs]: I would argue that you want to do an update ASAP. But if we decide to leave this as a cold observable, we can connect to it in the BasePersistState (it is in fact a side effect there, the return value is now thrown away).

**Comment by simonrozsival** [comment]
This is now ready for a review. The app has to be restarted after synchronization is complete to show the persisted data in the UI.

File: Toggl.Foundation/Sync/States/BasePersistState.cs [hunk 4]
```diff
@@ -25,7 +25,7 @@ public IObservable<ITransition> Start(FetchObservables fetch)
             return FetchObservable(fetch)
                 .SingleAsync()
                 .Select(entities => entities.Select(ConvertToDatabaseEntity).ToList())
-                .Do(databaseEntities => BatchUpdate(database, databaseEntities))
+                .Do(databaseEntities => BatchUpdate(database, databaseEntities).Wait())
```
Comment (amulware) [inline Toggl.Foundation/Sync/States/BasePersistState.cs]: Why wait sync with `.Wait()` and not async with `await`?

File: Toggl.Foundation/Sync/States/FetchAllSinceState.cs [hunk 5]
```diff
@@ -20,7 +21,8 @@ public FetchAllSinceState(ITogglDatabase database, ITogglApi api)
 
         public IObservable<ITransition> Start() => Observable.Create<ITransition>(observer =>
         {
-            var sinceDates = database.SinceParameters.Get();
+            var realmSinceDates = database.SinceParameters.Get();
```
Comment (amulware) [inline Toggl.Foundation/Sync/States/FetchAllSinceState.cs]: Don't include reference to realm. Foundation does not know what database is being used.

**Pull request merged**